command/select.go

```diff
@@ -27,94 +28,210 @@ Options:
-				Usage:   "SQL expression to use to select from the objects",
-			},
-			&cli.StringFlag{
-				Name:  "compression",
-				Usage: "input compression format",
-				Value: "NONE",
-			},
-			&cli.GenericFlag{
-				Name:  "format",
-				Usage: "input data format (only JSON supported for the moment)",
-				Value: &EnumValue{
-					Enum:    []string{"JSON"},
-					Default: "JSON",
+		Subcommands: []*cli.Command{
+			{
+				Name:  "csv",
+				Usage: "write queries for csv files",
+				Flags: append([]cli.Flag{
+					&cli.StringFlag{
+						Name:  "delimiter",
+						Usage: "delimiter of the csv file",
+						Value: ",",
+					},
+				}, sharedFlags...),
+				CustomHelpTemplate: defaultSelectHelpTemplate,
+				Before:             beforeFunc,
+				Action: func(c *cli.Context) (err error) {
+					delimiter := c.String("delimiter")
+					cmd, err := buildSelect(c, "csv", &delimiter)
+					if err != nil {
+						printError(cmd.fullCommand, c.Command.Name, err)
+						return err
+					}
+					return cmd.Run(c.Context)
 				},
 			},
-			&cli.StringSliceFlag{
-				Name:  "exclude",
-				Usage: "exclude objects with given pattern",
-			},
-			&cli.BoolFlag{
-				Name:  "force-glacier-transfer",
-				Usage: "force transfer of glacier objects whether they are restored or not",
-			},
-			&cli.BoolFlag{
-				Name:  "ignore-glacier-warnings",
-				Usage: "turns off glacier warnings: ignore errors encountered during selecting objects",
-			},
-			&cli.BoolFlag{
-				Name:  "raw",
-				Usage: "disable the wildcard operations, useful with filenames that contains glob characters",
-			},
-			&cli.BoolFlag{
-				Name:  "all-versions",
-				Usage: "list all versions of object(s)",
+			{
+				Name:  "json",
+				Usage: "write queries for json files",
+				Flags: append([]cli.Flag{
+					&cli.GenericFlag{
+						Name:  "structure",
+						Usage: "how objects are aligned in the json file",
+						Value: &EnumValue{
+							Enum:    []string{"lines", "document"},
+							Default: "lines",
+							ConditionFunction: func(str, target string) bool {
+								return strings.ToLower(target) == str
+							},
+						},
+					},
+				}, sharedFlags...),
+				CustomHelpTemplate: defaultSelectHelpTemplate,
+				Before:             beforeFunc,
+				Action: func(c *cli.Context) (err error) {
+					structure := c.String("structure")
+					cmd, err := buildSelect(c, "json", &structure)
+					if err != nil {
+						printError(cmd.fullCommand, c.Command.Name, err)
+						return err
+					}
+					return cmd.Run(c.Context)
+				},
 			},
-			&cli.StringFlag{
-				Name:  "version-id",
-				Usage: "use the specified version of the object",
+			{
+				Name:               "parquet",
+				Usage:              "write queries for parquet files",
+				Flags:              sharedFlags,
+				CustomHelpTemplate: defaultSelectHelpTemplate,
+				Before: func(c *cli.Context) (err error) {
+					if c.String("compression") != "" {
+						err = errors.New("compression is not supported for parquet files")
```

Comment:
I thought that in the future, if AWS starts to support compression we can simply remove the check, thus it would become a shared flag. But simply adding the compression flag in the future would be better I assume ?

Reply:
Resolved https://github.com/peak/s5cmd/pull/611/commits/126cb95f3e7676c3cd1869fbbde30edbe889b28c